cirq-google/cirq_google/devices/google_device.py

```diff
@@ -0,0 +1,226 @@
+# limitations under the License.
+
+"""Device object representing Google devices."""
+
+from typing import Any, Set, Tuple, cast
+import cirq
+from cirq_google.api import v2
+
+
+@cirq.value_equality
+class GoogleDevice(cirq.Device):
+    """Device object representing Google devices.
+
+    For end users, instances of this class are typically accessed via
+    `Engine.get_processor('processor_name').get_device()`.
+
+    This class is compliant with the core `cirq.Device` abstraction. In particular:
+        * Device information is captured in the `metadata` property.
+        * An instance of `GoogleDevice` can be used to validate circuits, moments, and operations.
+
+    Example use cases:
+
+        * Get an instance of a Google device.
+        >>> device = cirq_google.get_engine().get_processor('weber').get_device()
+
+        * Print the grid layout of the device.
+        >>> print(device)
+
+        * Determine whether a circuit can be run on the device.
+        >>> device.validate_circuit(circuit)  # Raises an exception if the circuit is invalid.
+
+        * Determine whether an operation can be run on the device.
+        >>> device.validate_operation(operation)  # Raises an exception if the operation is invalid.
+
+        * Get the `cirq.Gateset` containing valid gates for the device, and inspect the full list
+          of valid gates.
+        >>> gateset = device.metadata.gateset
+        >>> print(gateset)
+
+        * Determine whether a gate is available on the device.
+        >>> gate in device.metadata.gateset
+
+        * Get a collection of valid qubits on the device.
+        >>> device.metadata.qubit_set
+
+        * Get a collection of valid qubit pairs for two-qubit gates.
+        >>> device.metadata.qubit_pairs
+
+        * Get a collection of isolated qubits, i.e. qubits which are not part of any qubit pair.
+        >>> device.metadata.isolated_qubits
+
+        * Get a collection of approximate durations of performing each gate supported by the device.
+        >>> device.metadata.gate_durations
+
+        TODO(#5050) Add compilation_target_gatesets example.
+
+    Notes for cirq_google internal implementation:
+
+    For Google devices, the
+    [DeviceSpecification proto](
+        https://github.com/quantumlib/Cirq/blob/3969c2d3964cea56df33b329f036ba6810683882/cirq-google/cirq_google/api/v2/device.proto#L13
+    )
+    is the main specification for device information surfaced by the Quantum Computing Service.
+    Thus, this class is should be instantiated using a `DeviceSpecification` proto via the
+    `from_proto()` class method.
+    """
+
+    def __init__(self, metadata: cirq.GridDeviceMetadata):
+        """Creates a GoogleDevice object.
+
+        This constructor typically should not be used directly. Use `from_proto()` instead.
+        """
+        self._metadata = metadata
+
+    @classmethod
+    def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GoogleDevice':
```

Comment:
Why is the `from_proto` method preffered over init design-wise ?

Reply:
The full DeviceSpecification would have to be stored in the repr and json, which could potentially be verbose?